File: php/src/UIIDClient.php
<?php

namespace UIID\SDK;

use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;

class UIIDClient
{
    private string $baseUrl;
    private string $clientId;
    private string $clientSecret;
    private ?string $accessToken = null;
    private ClientInterface $client;

    public function __construct(
        string $clientId = '',
        string $clientSecret = '',
        string $baseUrl = 'https://uiid.linkspreed.com',
        ?ClientInterface $client = null
    ) {
        $this->clientId = $clientId;
        $this->clientSecret = $clientSecret;
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->client = $client ?? new Client(['base_uri' => $this->baseUrl]);
    }

    public function setAccessToken(string $token): void
    {
        $this->accessToken = $token;
    }

    public function getAccessToken(): ?string
    {
        return $this->accessToken;
    }

    private function request(string $method, string $endpoint, array $options = []): array
    {
        $headers = $options['headers'] ?? [];
        if ($this->accessToken && !isset($headers['Authorization'])) {
            $headers['Authorization'] = 'Bearer ' . $this->accessToken;
        }

        $options['headers'] = $headers;
        $options['http_errors'] = false;

        $response = $this->client->request($method, $endpoint, $options);
        $body = (string) $response->getBody();
        $data = json_decode($body, true) ?? ['raw' => $body];

        return [
            'status_code' => $response->getStatusCode(),
            'ok' => $response->getStatusCode() >= 200 && $response->getStatusCode() < 300,
            'data' => $data,
        ];
    }

    // --- 02. Authentication (OIDC) ---

    public function getAuthorizeUrl(string $redirectUri, string $scope = 'openid profile email alias:read:public'): string
    {
        $query = http_build_query([
            'client_id' => $this->clientId,
            'redirect_uri' => $redirectUri,
            'response_type' => 'code',
            'scope' => $scope,
        ]);

        return $this->baseUrl . '/oauth/authorize?' . $query;
    }

    public function exchangeCodeForToken(string $code, string $redirectUri): array
    {
        $res = $this->request('POST', '/oauth/token', [
            'form_params' => [
                'grant_type' => 'authorization_code',
                'code' => $code,
                'client_id' => $this->clientId,
                'client_secret' => $this->clientSecret,
                'redirect_uri' => $redirectUri,
            ],
        ]);

        if (isset($res['data']['access_token'])) {
            $this->accessToken = $res['data']['access_token'];
        }

        return $res['data'];
    }

    public function refreshToken(string $refreshToken): array
    {
        $res = $this->request('POST', '/oauth/token', [
            'form_params' => [
                'grant_type' => 'refresh_token',
                'refresh_token' => $refreshToken,
                'client_id' => $this->clientId,
                'client_secret' => $this->clientSecret,
            ],
        ]);

        if (isset($res['data']['access_token'])) {
            $this->accessToken = $res['data']['access_token'];
        }

        return $res['data'];
    }

    public function getUserInfo(): array
    {
        return $this->request('GET', '/oauth/userinfo')['data'];
    }

    public function getOidcDiscovery(): array
    {
        return $this->request('GET', '/.well-known/openid-configuration')['data'];
    }

    public function getJwks(): array
    {
        return $this->request('GET', '/.well-known/jwks.json')['data'];
    }

    // --- 03. App Management ---

    public function onboardApplication(string $name, string $redirectUri): array
    {
        return $this->request('POST', '/api/v1/applications', [
            'json' => ['name' => $name, 'redirect_uri' => $redirectUri],
        ])['data'];
    }

    public function revokeApplication(string $appId): array
    {
        return $this->request('DELETE', "/api/v1/applications/{$appId}")['data'];
    }

    // --- 04. Core ID API ---

    public function generateCoreID(): array
    {
        return $this->request('GET', '/api/v1/core/uiid/generate')['data'];
    }

    public function getKycStatus(): array
    {
        return $this->request('GET', '/api/v1/core/kyc/status')['data'];
    }

    public function storeCoreData(string $key, mixed $value, bool $isPublic = false, bool $isImmutable = false): array
    {
        return $this->request('POST', '/api/v1/core/data', [
            'json' => [
                'key' => $key,
                'value' => $value,
                'is_public' => $isPublic,
                'is_immutable' => $isImmutable,
            ],
        ])['data'];
    }

    public function getAuthorizedApplications(): array
    {
        return $this->request('GET', '/api/v1/core/applications')['data'];
    }

    // --- 05. Alias Network ---

    public function getAliases(): array
    {
        return $this->request('GET', '/api/v1/aliases')['data'];
    }

    public function createAlias(string $name): array
    {
        return $this->request('POST', '/api/v1/aliases/create', [
            'json' => ['name' => $name],
        ])['data'];
    }

    public function updateAliasStatus(string $aliasId, string $status, ?string $secretKey = null): array
    {
        $body = ['status' => $status];
        if ($secretKey) {
            $body['key'] = $secretKey;
        }

        return $this->request('PUT', "/api/v1/aliases/{$aliasId}/status", [
            'json' => $body,
        ])['data'];
    }

    public function getAliasData(string $aliasId): array
    {
        return $this->request('GET', "/api/v1/aliases/data/{$aliasId}")['data'];
    }

    public function overwriteAliasData(string $aliasId, array $data): array
    {
        return $this->request('POST', '/api/v1/aliases/data', [
            'json' => ['alias_id' => $aliasId, 'data' => $data],
        ])['data'];
    }

    public function patchAliasData(array $payload): array
    {
        return $this->request('PATCH', '/api/v1/aliases/data', [
            'json' => $payload,
        ])['data'];
    }

    public function requestNodeDeletion(string $aliasId, string $key): array
    {
        return $this->request('POST', '/api/v1/aliases/storage/request-deletion', [
            'json' => ['alias_id' => $aliasId, 'key' => $key],
        ])['data'];
    }

    public function removeAliasKey(string $aliasId, string $key): array
    {
        return $this->request('DELETE', '/api/v1/aliases/data', [
            'json' => ['alias_id' => $aliasId, 'key' => $key],
        ])['data'];
    }

    public function purgeAlias(string $aliasId): array
    {
        return $this->request('DELETE', "/api/v1/aliases/{$aliasId}")['data'];
    }

    // --- Shared Buckets (v2.6) ---

    public function addAliasMember(string $aliasIdStr, string $targetUiid, string $role): array
    {
        return $this->request('POST', '/api/v1/aliases/members', [
            'json' => [
                'alias_id_str' => $aliasIdStr,
                'target_uiid' => $targetUiid,
                'role' => $role,
            ],
        ])['data'];
    }

    public function getAliasMembers(string $aliasId): array
    {
        return $this->request('GET', "/api/v1/aliases/members/{$aliasId}")['data'];
    }

    // --- 06. Ecosystem & Trust ---

    public function getMarketplaceApps(): array
    {
        return $this->request('GET', '/api/v1/marketplace/apps')['data'];
    }

    public function getMarketplaceAppDetails($appId): array
    {
        return $this->request('GET', "/api/v1/marketplace/apps/{$appId}")['data'];
    }

    public function getBadges(): array
    {
        return $this->request('GET', '/api/v1/badges')['data'];
    }

    public function getUserBadges(): array
    {
        return $this->request('GET', '/api/v1/user/badges')['data'];
    }

    // --- 07. Auditing ---

    public function getCoreAuditLogs(): array
    {
        return $this->request('GET', '/api/v1/audit/core')['data'];
    }

    public function getAliasAuditLogs(string $aliasId): array
    {
        return $this->request('GET', "/api/v1/audit/alias?alias_id={$aliasId}")['data'];
    }

    // --- 08. Local Integration ---

    public function checkAuthStatus(): array
    {
        return $this->request('GET', '/api/v1/auth/check')['data'];
    }

    // --- 10. Webhooks ---

    public function subscribeWebhook(string $url, array $events, ?string $aliasId = null): array
    {
        $body = ['url' => $url, 'events' => $events];
        if ($aliasId) {
            $body['alias_id'] = $aliasId;
        }

        return $this->request('POST', '/api/v1/webhooks', [
            'json' => $body,
        ])['data'];
    }

    public static function verifyWebhookSignature(array|string $payload, string $signature, string $secret): bool
    {
        $bodyStr = is_array($payload) ? json_encode($payload, JSON_UNESCAPED_SLASHES) : $payload;
        $computed = hash_hmac('sha256', $bodyStr, $secret);
        return hash_equals($signature, $computed);
    }
}

File: python/uiid_client.py
import hmac
import hashlib
import json
from urllib.parse import urlencode, quote
import requests

class UIIDClient:
    def __init__(self, client_id="", client_secret="", base_url="https://uiid.linkspreed.com", session=None):
        self.client_id = client_id
        self.client_secret = client_secret
        self.base_url = base_url.rstrip("/")
        self.access_token = None
        self.session = session or requests.Session()

    def set_access_token(self, token: str):
        self.access_token = token

    def _request(self, method: str, endpoint: str, params=None, data=None, json_data=None, headers=None):
        url = endpoint if endpoint.startswith("http") else f"{self.base_url}{endpoint}"
        req_headers = headers or {}

        if self.access_token and "Authorization" not in req_headers:
            req_headers["Authorization"] = f"Bearer {self.access_token}"

        response = self.session.request(
            method=method,
            url=url,
            params=params,
            data=data,
            json=json_data,
            headers=req_headers
        )
        try:
            res_data = response.json()
        except Exception:
            res_data = {"text": response.text}

        return {
            "status_code": response.status_code,
            "ok": response.ok,
            "data": res_data
        }

    # --- 02. Authentication (OIDC) ---

    def get_authorize_url(self, redirect_uri: str, scope: str = "openid profile email alias:read:public") -> str:
        params = {
            "client_id": self.client_id,
            "redirect_uri": redirect_uri,
            "response_type": "code",
            "scope": scope
        }
        return f"{self.base_url}/oauth/authorize?{urlencode(params)}"

    def exchange_code_for_token(self, code: str, redirect_uri: str) -> dict:
        payload = {
            "grant_type": "authorization_code",
            "code": code,
            "client_id": self.client_id,
            "client_secret": self.client_secret,
            "redirect_uri": redirect_uri
        }
        res = self._request("POST", "/oauth/token", data=payload)
        if res.get("data") and "access_token" in res["data"]:
            self.access_token = res["data"]["access_token"]
        return res["data"]

    def refresh_token(self, refresh_token: str) -> dict:
        payload = {
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
            "client_id": self.client_id,
            "client_secret": self.client_secret
        }
        res = self._request("POST", "/oauth/token", data=payload)
        if res.get("data") and "access_token" in res["data"]:
            self.access_token = res["data"]["access_token"]
        return res["data"]

    def get_user_info(self) -> dict:
        res = self._request("GET", "/oauth/userinfo")
        return res["data"]

    def get_oidc_discovery(self) -> dict:
        res = self._request("GET", "/.well-known/openid-configuration")
        return res["data"]

    def get_jwks(self) -> dict:
        res = self._request("GET", "/.well-known/jwks.json")
        return res["data"]

    # --- 03. App Management ---

    def onboard_application(self, name: str, redirect_uri: str) -> dict:
        res = self._request("POST", "/api/v1/applications", json_data={"name": name, "redirect_uri": redirect_uri})
        return res["data"]

    def revoke_application(self, app_id: str) -> dict:
        res = self._request("DELETE", f"/api/v1/applications/{app_id}")
        return res["data"]

    # --- 04. Core ID API ---

    def generate_core_id(self) -> dict:
        res = self._request("GET", "/api/v1/core/uiid/generate")
        return res["data"]

    def get_kyc_status(self) -> dict:
        res = self._request("GET", "/api/v1/core/kyc/status")
        return res["data"]

    def store_core_data(self, key: str, value, is_public: bool = False, is_immutable: bool = False) -> dict:
        payload = {
            "key": key,
            "value": value,
            "is_public": is_public,
            "is_immutable": is_immutable
        }
        res = self._request("POST", "/api/v1/core/data", json_data=payload)
        return res["data"]

    def get_authorized_applications(self) -> dict:
        res = self._request("GET", "/api/v1/core/applications")
        return res["data"]

    # --- 05. Alias Network ---

    def get_aliases(self) -> dict:
        res = self._request("GET", "/api/v1/aliases")
        return res["data"]

    def create_alias(self, name: str) -> dict:
        res = self._request("POST", "/api/v1/aliases/create", json_data={"name": name})
        return res["data"]

    def update_alias_status(self, alias_id: str, status: str, secret_key: str = None) -> dict:
        payload = {"status": status}
        if secret_key:
            payload["key"] = secret_key
        res = self._request("PUT", f"/api/v1/aliases/{alias_id}/status", json_data=payload)
        return res["data"]

    def get_alias_data(self, alias_id: str) -> dict:
        res = self._request("GET", f"/api/v1/aliases/data/{alias_id}")
        return res["data"]

    def overwrite_alias_data(self, alias_id: str, data: dict) -> dict:
        res = self._request("POST", "/api/v1/aliases/data", json_data={"alias_id": alias_id, "data": data})
        return res["data"]

    def patch_alias_data(self, payload: dict) -> dict:
        res = self._request("PATCH", "/api/v1/aliases/data", json_data=payload)
        return res["data"]

    def request_node_deletion(self, alias_id: str, key: str) -> dict:
        res = self._request("POST", "/api/v1/aliases/storage/request-deletion", json_data={"alias_id": alias_id, "key": key})
        return res["data"]

    def remove_alias_key(self, alias_id: str, key: str) -> dict:
        res = self._request("DELETE", "/api/v1/aliases/data", json_data={"alias_id": alias_id, "key": key})
        return res["data"]

    def purge_alias(self, alias_id: str) -> dict:
        res = self._request("DELETE", f"/api/v1/aliases/{alias_id}")
        return res["data"]

    # --- Shared Buckets (v2.6) ---

    def add_alias_member(self, alias_id_str: str, target_uiid: str, role: str) -> dict:
        payload = {"alias_id_str": alias_id_str, "target_uiid": target_uiid, "role": role}
        res = self._request("POST", "/api/v1/aliases/members", json_data=payload)
        return res["data"]

    def get_alias_members(self, alias_id: str) -> dict:
        res = self._request("GET", f"/api/v1/aliases/members/{alias_id}")
        return res["data"]

    # --- 06. Ecosystem & Trust ---

    def get_marketplace_apps(self) -> dict:
        res = self._request("GET", "/api/v1/marketplace/apps")
        return res["data"]

    def get_marketplace_app_details(self, app_id) -> dict:
        res = self._request("GET", f"/api/v1/marketplace/apps/{app_id}")
        return res["data"]

    def get_badges(self) -> dict:
        res = self._request("GET", "/api/v1/badges")
        return res["data"]

    def get_user_badges(self) -> dict:
        res = self._request("GET", "/api/v1/user/badges")
        return res["data"]

    # --- 07. Auditing ---

    def get_core_audit_logs(self) -> dict:
        res = self._request("GET", "/api/v1/audit/core")
        return res["data"]

    def get_alias_audit_logs(self, alias_id: str) -> dict:
        res = self._request("GET", f"/api/v1/audit/alias?alias_id={alias_id}")
        return res["data"]

    # --- 08. Local Integration & Smart Tags ---

    def check_auth_status(self) -> dict:
        res = self._request("GET", "/api/v1/auth/check")
        return res["data"]

    # --- 10. Webhooks ---

    def subscribe_webhook(self, url: str, events: list, alias_id: str = None) -> dict:
        payload = {"url": url, "events": events}
        if alias_id:
            payload["alias_id"] = alias_id
        res = self._request("POST", "/api/v1/webhooks", json_data=payload)
        return res["data"]

    @staticmethod
    def verify_webhook_signature(payload, signature: str, secret: str) -> bool:
        body_bytes = json.dumps(payload, separators=(',', ':')).encode('utf-8') if isinstance(payload, dict) else str(payload).encode('utf-8')
        computed = hmac.new(secret.encode('utf-8'), body_bytes, hashlib.sha256).hexdigest()
        return hmac.compare_digest(signature, computed)
